src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,49 +1,118 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
 
+    public static final int DEFAULT_CAPACITY = 10;
+
+    private T[] values;
+    private int size;
+
+    public ArrayList() {
+        values = (T[]) new Object[DEFAULT_CAPACITY];
+        size = 0;
+    }
+
     @Override
     public void add(T value) {
-
+        if (!isSpaceAvailable()) {
+            grow();
+        }
+        values[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (!isSpaceAvailable()) {
+            grow();
+        }
 
+        if (isIndexInvalid(index) || index > size) {
+            throw new ArrayIndexOutOfBoundsException();
+        }
+
+        System.arraycopy(values, index, values, index + 1, size - index);
+        values[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (isIndexInvalid(index)) {
+            throw new ArrayIndexOutOfBoundsException();
+        }
+
+        if (index < size) {
+            return values[index];
+        }
+        throw new ArrayIndexOutOfBoundsException();
     }
 
     @Override
     public void set(T value, int index) {
+        if (isIndexInvalid(index)) {
```

Comment:
let's add another condition here as well

so that we could simplify this method and use only one `if` and one `throw`